File: elixir/beer-song/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(number) do
    line_one(number) <> "\n" <> line_two(number) <> "\n"
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0) do
    range
    |> Stream.map(&verse/1)
    |> Enum.join("\n")
  end

  defp beers(0), do: beers("no more")
  defp beers(1), do: "1 bottle of beer"
  defp beers(n), do: "#{n} bottles of beer"

  defp line_one(0),
    do: line_one("no more")
  defp line_one(n),
    do: "#{n |> beers |> String.capitalize} on the wall, #{beers(n)}."

  defp line_two(0),
    do: "Go to the store and buy some more, #{beers(99)} on the wall."
  defp line_two(1),
    do: "Take it down and pass it around, #{beers(0)} on the wall."
  defp line_two(n),
    do: "Take one down and pass it around, #{beers(n - 1)} on the wall."
end

File: c/beer-song/beer_song.c
#include "beer_song.h"
#include <stdint.h>
#include <stdio.h>

void write_verse(const int8_t bottles, const int8_t index, char **song);
void write_line_one(const int8_t bottles, char *song);
void write_line_two(const int8_t bottles, char *song);

void write_verse(const int8_t bottles, const int8_t index, char **song) {
  write_line_one(bottles, song[index]);
  write_line_two(bottles - 1, song[index + 1]);
}

void write_line_one(const int8_t bottles, char *song) {
  if (bottles == 0) {
    sprintf(song,
            "No more bottles of beer on the wall, no more bottles of beer.");
  } else if (bottles == 1) {
    sprintf(song, "1 bottle of beer on the wall, 1 bottle of beer.");
  } else {
    const char *fmt = "%u bottles of beer on the wall, %u bottles of beer.";
    int size = snprintf(NULL, 0, fmt, bottles, bottles);
    snprintf(song, size + 1, fmt, bottles, bottles);
  }
}
void write_line_two(const int8_t bottles, char *song) {
  if (bottles == -1) {
    sprintf(
        song,
        "Go to the store and buy some more, 99 bottles of beer on the wall.");
  } else if (bottles == 0) {
    sprintf(song, "Take it down and pass it around, no more bottles of beer "
                  "on the wall.");
  } else if (bottles == 1) {
    sprintf(song, "Take one down and pass it around, 1 bottle of beer "
                  "on the wall.");
  } else {
    const char *fmt =
        "Take one down and pass it around, %u bottles of beer on the wall.";
    int size = snprintf(NULL, 0, fmt, bottles);
    snprintf(song, size + 1, fmt, bottles);
  }
}

void recite(const int8_t start_bottles, const int8_t take_down, char **song) {
  int8_t number_of_bottles = start_bottles;
  for (int8_t i = 0; i < take_down; ++i) {
    write_verse(number_of_bottles--, i * 3, song);
  }
}
